src/Cameras/Inputs/BaseCameraPointersInput.ts

```diff
@@ -0,0 +1,355 @@
+                else if (pointA && pointB) {
+                    var ed = (pointA.pointerId === evt.pointerId) ? pointA : pointB;
+                    ed.x = evt.clientX;
+                    ed.y = evt.clientY;
+                    var distX = pointA.x - pointB.x;
+                    var distY = pointA.y - pointB.y;
+                    var pinchSquaredDistance = (distX * distX) + (distY * distY);
+                    var multiTouchPanPosition = {x: (pointA.x + pointB.x) / 2,
+                                                 y: (pointA.y + pointB.y) / 2,
+                                                 pointerId: evt.pointerId,
+                                                 type: p.type};
+
+                    this.onMultiTouch(
+                      pointA,
+                      pointB,
+                      previousPinchSquaredDistance,
+                      pinchSquaredDistance,
+                      previousMultiTouchPanPosition,
+                      multiTouchPanPosition);
+
+                    previousMultiTouchPanPosition = multiTouchPanPosition;
+                    previousPinchSquaredDistance = pinchSquaredDistance;
+                }
+            }
+        };
+
+        this._observer = this.camera.getScene().onPointerObservable.add(
+            this._pointerInput,
+            PointerEventTypes.POINTERDOWN | PointerEventTypes.POINTERUP |
+            PointerEventTypes.POINTERMOVE);
+
+        this._onLostFocus = () => {
+            pointA = pointB = null;
+            previousPinchSquaredDistance = 0;
+            previousMultiTouchPanPosition = null;
+            this.onLostFocus();
+        };
+
+        element.addEventListener("contextmenu",
+            <EventListener>this.onContextMenu.bind(this), false);
+
+        Tools.RegisterTopRootEvents([
+            { name: "blur", handler: this._onLostFocus }
+        ]);
+    }
+
+    /**
+     * Detach the current controls from the specified dom element.
+     * @param element Defines the element to stop listening the inputs from
+     */
+    public detachControl(element: Nullable<HTMLElement>): void {
+        if (this._onLostFocus) {
+            Tools.UnregisterTopRootEvents([
+                { name: "blur", handler: this._onLostFocus }
+            ]);
+        }
+
+        if (element && this._observer) {
+            this.camera.getScene().onPointerObservable.remove(this._observer);
+            this._observer = null;
+
+            if (this.onContextMenu) {
+                element.removeEventListener("contextmenu", <EventListener>this.onContextMenu);
+            }
+
+            this._onLostFocus = null;
+        }
+
+        this._altKey = false;
+        this._ctrlKey = false;
+        this._metaKey = false;
+        this._shiftKey = false;
+        this._buttonsPressed = 0;
+    }
+
+    /**
+     * Gets the class name of the current input.
+     * @returns the class name
+     */
+    public getClassName(): string {
```

Comment:
why not overloading the function on each child?

Reply:
> why not overloading the function on each child?

I already mentioned this in the "conversation" tab ~17 days ago...
..."need to be fixed with correct name"...

While overloading the function certainly looks like the correct thing to do here, it's not actually that simple.
`getClassName()` is called from the constructor of one of the parent classes. That seems to mean it does not exist in `this` of the child classes until after the constructor is called....

This is all new to me so it's possible/likely my interpretation is wrong; i'm not a JS guy.

I'm afraid my brain has dropped the exact error... i'll recreate when i get home this evening if what i'm describing here makes no sense.

Reply:
well this is the case on all our classes:
* https://github.com/BabylonJS/Babylon.js/blob/master/src/Meshes/linesMesh.ts#L103
* https://github.com/BabylonJS/Babylon.js/blob/master/src/Meshes/groundMesh.ts#L44

Reply:
sure thing:)

Reply:
Hrm, interesting...
so i originally wanted `getClassName()` in the base class to be virtual/abstract to force child implementations to set it.
Unfortunately setting `abstract public getClassName(): string;` in the base class causes the behavior i described; The method is undefined at compile time so the code errors at run time.

Setting `public getClassName(): string { return "something"; }` in the base class then overriding it later works; There is always a valid implementation, even if the one in Base should never be used.

I've submitted a fix that compiles and runs cleanly.
There is an argument that we should put a `console.error()` in `BaseCameraPointersInput.getClassName()`.

Thoughts?

Reply:
Nope. marking it abstract breaks at run time.
Let's leave it as it is.